app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt

```diff
@@ -453,6 +461,24 @@ class StatePlayerRecyclerViewAssembler private constructor(
     animateCongratulationsTextView(textView)
   }
 
+  /** Shows confetti when the learner reaches the end of an exploration. */
+  private fun maybeShowCelebrationForEndOfExplorationSession(isTablet: Boolean) {
+    check(playerFeatureSet.showCelebrationAtEndOfExplorationSession) {
+      "Cannot show end of exploration confetti for assembler that doesn't support it"
+    }
+    val confettiView = checkNotNull(fullScreenConfettiView) {
+      "Expected non-null reference to confetti view"
+    }
+    val colorsList = checkNotNull(confettiColors) {
+      "Expected non-null list of confetti colors"
+    }
+    if (!confettiView.isActive()) {
+      // If learners toggle back and forth from the ene dof the exploration we only show the confetti one
+      // instance at a time.
```

Comment:
Just to confirm: is there a test to verify this?

Reply:
Yes, `testStateFragment_reachEndOfExplorationTwice_endOfSessionConfettiIsDisplayedOnce`